### ISSUE TITLE
Wrong auto import paths for entry points

### ISSUE THREAD

**Issue by cyrilletuzi** [Microsoft/TypeScript#23011]
**TypeScript Version:**  2.9.0-dev.20180330

**Search Terms:** imports, index

**Code**

With the following very common architecture : 
```
src/
|- external.ts
|- feature/
    |- feature.ts
    |- index.ts
    |- internal.ts
````

with `feature.ts` :
```ts
export class Feature {}
```

and `index.ts` as an entry point for this feature :
```ts
export { Feature } from './feature';
```

Then, try to auto-import `Feature` from `external.ts` and from `internal.ts`.

**Expected behavior:**

From `external.ts`, it should :
```ts
import { Feature } from "./feature/index";
```
(without `index` if `moduleResolution` is set to `node`).

From `internal.ts`, it should :
```ts
import { Feature } from "./feature";
```

**Actual behavior:**

From `external.ts`, it's OK :
```ts
import { Feature } from "./feature/index";
```

From `internal.ts`, it's also importing from the entry point, instead of directly from the file :
```ts
import { Feature } from "./index";
```
worse with `moduleResolution` set to `node`, as it will do this :
```ts
import { Feature } from ".";
```

**Related Issues:**
The current behavior was introduced for the opposite issue : for libs, imports must be from entry points and not from deep imports (otherwise it causes errors). So for now, TypeScript always choose the shortest path.

But inside a lib/feature, imports must be done from deep imports and not from entry points. Otherwise, it really often create circular dependencies issues (which, by the way, are really difficult to debug).

TypeScript should be smarter and do auto-imports as follow :
- for libs (`node_modules` and TS custom `paths`), choose the shorter path ;
- inside the project, if the file we're in is in the same directory (like `internal.ts`) or a subdirectory than the entry point (`index.ts`), choose the deep import ;
- otherwise, choose the shortest path.

**Comment by andy-ms** [comment]
> From internal.ts, it should : import { Feature } from "./feature/feature";`

But you're in `feature/internal` and that's a relative path, so wouldn't it be accessing `feature/feature/feature` ?

**Comment by cyrilletuzi** [comment]
It was just a typo, edited.

**Comment by cyrilletuzi** [comment]
I wouldn't say it's just a suggestion. In many cases it produces broken code that won't work, so for me it's a bug.

**Comment by andy-ms** [comment]
> it produces broken code that won't work

What's the failure you're getting?

**Comment by cyrilletuzi** [comment]
It's not a compilation failure, TS will be OK to compile it, but then there will be execution failures, which completely depends on the code you're doing.

The general problem is that importing from the entry point when you're inside a lib/feature really often create circular dependencies errors.

**Comment by cyrilletuzi** [comment]
What feedback do you need to take this issue seriously? I don't understand why such issues, which are not special cases at all but very common use cases and day to day development are just ignored.

**Comment by andy-ms** [comment]
Should be fixed by #23623
